lib/services/local/access_list.go

```diff
@@ -271,10 +273,49 @@ func (a *AccessListService) ListAccessListMembers(ctx context.Context, accessLis
 }
 
 // ListAllAccessListMembers returns a paginated list of all access list members for all access lists.
-func (a *AccessListService) ListAllAccessListMembers(ctx context.Context, pageSize int, pageToken string) (members []*accesslist.AccessListMember, nextToken string, err error) {
+func (a *AccessListService) ListAllAccessListMembers(ctx context.Context, pageSize int, pageToken string) ([]*accesslist.AccessListMember, string, error) {
 	// Locks are not used here as these operations are more likely to be used by the cache.
-	// Lists all access list members for all access lists.
-	return a.memberService.ListResources(ctx, pageSize, nextToken)
+	rangeStart := backend.Key(accessListMemberPrefix, pageToken)
+	rangeEnd := backend.RangeEnd(backend.ExactKey(accessListMemberPrefix))
+
+	// Adjust page size, so it can't be too large.
+	if pageSize <= 0 || pageSize > int(accessListMemberMaxPageSize) {
+		pageSize = int(accessListMemberMaxPageSize)
+	}
+
+	limit := pageSize + 1
+
+	// no filter provided get the range directly
+	result, err := a.backend.GetRange(ctx, rangeStart, rangeEnd, limit)
+	if err != nil {
+		return nil, "", trace.Wrap(err)
+	}
+
+	out := make([]*accesslist.AccessListMember, 0, len(result.Items))
+	for _, item := range result.Items {
+		resource, err := services.UnmarshalAccessListMember(item.Value, services.WithRevision(item.Revision), services.WithResourceID(item.ID))
+		if err != nil {
+			return nil, "", trace.Wrap(err)
+		}
+		out = append(out, resource)
+	}
+
+	var nextKey string
+	if len(out) > pageSize {
+		// For listing all access list members for all access lists, we want to list members regardless of which
+		// access list they belong to. As a result, we'll use custom pagination logic here to include the appropriate
+		// prefix across pages.
+		lastItemKey := result.Items[len(out)-1].Key
+		prefixOffset := 2 + len(accessListMemberPrefix)
+		if len(lastItemKey) < prefixOffset {
+			return nil, "", trace.BadParameter("unable to calculate next page for listing all access list members")
+		}
+		nextKey = string(lastItemKey[prefixOffset:])
```

Comment:
Here and in `ListAllAccessListReviews`: the logic should be the same as the one used to pick the key for upsert, it's just a bit hidden on upsert because of the `WithPrefix` service wrapper.
```suggestion
		nextKey = out[pageSize].Spec.AccessList + string(backend.Separator) + out[pageSize].Metadata.Name
```

Reply:
Maybe the generic service could grow an extra optional `func(T) string` to get the intended backend key to append to the prefix and to make pagination keys from, rather than relying on the simple heuristic that prepends a host id if the resource has a `GetHostID() string`?

Reply:
> Can we drop all of the custom listing here if we increase the page size by one and then create the true next key from the last access list member or access list review?

This doesn't work because the generic won't let pages exceed the max size set at initialization.

Reply:
I've adjusted this per our Slack conversation and introduced a `ListResourcesReturnNextResource`. Let me know what you think.